File: logfalcon/fc/detector.py
"""FC detection and identification via MSP handshake."""

from __future__ import annotations

import logging
from dataclasses import dataclass

from logfalcon.msp.client import MSPClient, MSPError
from logfalcon.msp.constants import (
    BLACKBOX_DEVICE_FLASH,
    BLACKBOX_DEVICE_NONE,
    BLACKBOX_DEVICE_SDCARD,
    BTFL_VARIANT,
    SUPPORTED_VARIANTS,
)

log = logging.getLogger(__name__)


@dataclass
class FCInfo:
    api_major: int
    api_minor: int
    variant: bytes  # e.g. b'BTFL' or b'INAV'
    uid: str  # hex string, e.g. "12ab34cdef..."
    blackbox_device: int


class FCDetectionError(Exception):
    pass


class FCNotSupported(FCDetectionError):
    """FC variant is not supported (not Betaflight or iNav)."""

    pass


# Keep old name as alias for backwards compatibility
FCNotBetaflight = FCNotSupported


class FCSDCardBlackbox(FCDetectionError):
    """FC uses SD card for blackbox — must be read directly."""

    pass


class FCBlackboxEmpty(FCDetectionError):
    """Flash is already empty — nothing to sync."""

    pass


def detect_fc(client: MSPClient) -> FCInfo:
    """Run MSP handshake, verify supported FC variant, return FC info.

    Raises:
        FCNotSupported: Variant not in SUPPORTED_VARIANTS
        FCSDCardBlackbox: Blackbox device is SD card
        FCDetectionError: Other identification failure
    """
    try:
        major, minor = client.get_api_version()
        log.info('MSP API version: %d.%d', major, minor)
    except MSPError as exc:
        raise FCDetectionError(f'MSP API_VERSION failed: {exc}') from exc

    try:
        variant = client.get_fc_variant()
        log.info('FC variant: %r', variant)
    except MSPError as exc:
        raise FCDetectionError(f'MSP FC_VARIANT failed: {exc}') from exc

    if variant[:4] not in SUPPORTED_VARIANTS:
        raise FCNotSupported(
            f'Unsupported FC variant {variant!r} — expected one of: '
            + ', '.join(v.decode() for v in sorted(SUPPORTED_VARIANTS))
        )

    uid = 'unknown'
    try:
        uid = client.get_uid()
        log.info('FC UID: %s', uid)
    except MSPError:
        log.warning("Could not read FC UID, using 'unknown'")

    # iNav deprecated MSP_BLACKBOX_CONFIG (80) — returns all zeros.
    # For iNav, skip the device-type check and infer flash support from
    # DATAFLASH_SUMMARY later in the orchestrator.
    blackbox_device = BLACKBOX_DEVICE_NONE
    if variant[:4] == BTFL_VARIANT:
        try:
            bb_cfg = client.get_blackbox_config()
            blackbox_device = bb_cfg.get('device', BLACKBOX_DEVICE_NONE)
            log.info('Blackbox device type: %d', blackbox_device)
        except MSPError as exc:
            log.warning('Could not read BLACKBOX_CONFIG: %s', exc)
    else:
        # For iNav, assume flash until DATAFLASH_SUMMARY proves otherwise
        blackbox_device = BLACKBOX_DEVICE_FLASH
        log.info('Non-Betaflight FC — skipping BLACKBOX_CONFIG, assuming flash')

    if blackbox_device == BLACKBOX_DEVICE_SDCARD:
        raise FCSDCardBlackbox(
            'FC uses SD card for blackbox — remove the FC SD card and read it directly'
        )

    return FCInfo(
        api_major=major,
        api_minor=minor,
        variant=variant,
        uid=uid,
        blackbox_device=blackbox_device,
    )

File: internal/fc/detector.go
// Package fc provides flight-controller detection via MSP handshake.
package fc

import (
	"fmt"
	"log/slog"

	"github.com/proeugene/logfalcon/internal/msp"
)

// FCInfo holds identification info from the MSP handshake.
type FCInfo struct {
	APIMajor        int
	APIMinor        int
	FirmwareVersion string // e.g. "4.5.0" — empty if MSPFCVersion not supported
	Variant         string // "BTFL" or "INAV"
	UID             string // hex string
	BlackboxDevice  int
	Warning         string // non-empty when firmware is newer than max tested
}

// DetectionError indicates a generic MSP identification failure.
type DetectionError struct {
	Message string
}

func (e *DetectionError) Error() string { return e.Message }

// NotSupportedError indicates the FC variant is not Betaflight or iNav.
type NotSupportedError struct {
	Variant string
}

func (e *NotSupportedError) Error() string {
	return fmt.Sprintf("unsupported FC variant %q", e.Variant)
}

// SDCardError indicates the FC uses an SD card for blackbox storage.
type SDCardError struct{}

func (e *SDCardError) Error() string {
	return "FC uses SD card for blackbox — remove the FC SD card and read it directly"
}

// BlackboxEmptyError indicates the flash is already empty.
type BlackboxEmptyError struct{}

func (e *BlackboxEmptyError) Error() string {
	return "flash is already empty — nothing to sync"
}

// MSPClient is the interface the detector needs from the MSP client.
// Using an interface allows testing without a real serial port.
type MSPClient interface {
	GetAPIVersion() (int, int, error)
	GetFCVersion() (string, error)
	GetFCVariant() (string, error)
	GetUID() (string, error)
	GetBlackboxConfig() (int, error)
}

// Detect runs the MSP handshake and returns FC info.
//
// Steps:
//  1. GetAPIVersion — log result; wrap errors as DetectionError
//  2. GetFCVariant  — check against SupportedVariants; NotSupportedError if unknown
//  3. GetFCVersion  — human-readable firmware version string (best effort)
//  4. CheckVersion  — VersionTooOldError = hard stop; VersionTooNewError = warning
//  5. GetUID        — use "unknown" on error
//  6. GetBlackboxConfig — BTFL queries MSP; INAV skips (assumes flash)
//  7. SDCard device → SDCardError
func Detect(client MSPClient) (*FCInfo, error) {
	// 1. API version
	major, minor, err := client.GetAPIVersion()
	if err != nil {
		return nil, &DetectionError{Message: fmt.Sprintf("MSP API_VERSION failed: %v", err)}
	}
	slog.Info("MSP API version", "major", major, "minor", minor)

	// 2. FC variant
	variant, err := client.GetFCVariant()
	if err != nil {
		return nil, &DetectionError{Message: fmt.Sprintf("MSP FC_VARIANT failed: %v", err)}
	}
	slog.Info("FC variant", "variant", variant)

	if len(variant) > 4 {
		variant = variant[:4]
	}
	if !msp.SupportedVariants[variant] {
		return nil, &NotSupportedError{Variant: variant}
	}

	// 3. Firmware version string (best effort — some older FC builds don't support this)
	firmwareVersion := ""
	if fv, err := client.GetFCVersion(); err == nil {
		firmwareVersion = fv
		slog.Info("FC firmware version", "version", firmwareVersion)
	} else {
		slog.Debug("MSPFCVersion not available", "error", err)
	}

	// 4. Version check — block if too old, warn if too new
	warning := ""
	if verr := CheckVersion(variant, firmwareVersion, major, minor); verr != nil {
		switch verr.(type) {
		case *VersionTooOldError:
			return nil, verr
		case *VersionTooNewError:
			warning = verr.Error()
			slog.Warn("FC firmware newer than tested", "warning", warning)
		}
	}

	// 5. UID — best effort
	uid := "unknown"
	if u, err := client.GetUID(); err == nil {
		uid = u
		slog.Info("FC UID", "uid", uid)
	} else {
		slog.Warn("could not read FC UID, using 'unknown'")
	}

	// 6. Blackbox config
	bbDevice := msp.BlackboxDeviceNone
	if variant == msp.BTFLVariant {
		deviceType, err := client.GetBlackboxConfig()
		if err != nil {
			slog.Warn("could not read BLACKBOX_CONFIG", "error", err)
		} else {
			bbDevice = deviceType
			slog.Info("blackbox device type", "device", bbDevice)
		}
	} else {
		// iNav deprecated MSP_BLACKBOX_CONFIG — assume flash until
		// DATAFLASH_SUMMARY proves otherwise.
		bbDevice = msp.BlackboxDeviceFlash
		slog.Info("non-Betaflight FC — skipping BLACKBOX_CONFIG, assuming flash")
	}

	// 7. SD card → error
	if bbDevice == msp.BlackboxDeviceSDCard {
		return nil, &SDCardError{}
	}

	return &FCInfo{
		APIMajor:        major,
		APIMinor:        minor,
		FirmwareVersion: firmwareVersion,
		Variant:         variant,
		UID:             uid,
		BlackboxDevice:  bbDevice,
		Warning:         warning,
	}, nil
}
